src/components/PVGenerator/PVGeneratorComponent.tsx

```diff
@@ -0,0 +1,233 @@
+import * as React from "react";
+import {action, computed, makeObservable, observable} from "mobx";
+import {observer} from "mobx-react";
+import {FormGroup, Tabs, Tab, TabId, HTMLSelect, AnchorButton, Position} from "@blueprintjs/core";
+import {Tooltip2} from "@blueprintjs/popover2";
+import ReactResizeDetector from "react-resize-detector";
+import {AppStore, DefaultWidgetConfig, HelpType, WidgetProps, WidgetsStore} from "stores";
+import {PVGeneratorWidgetStore, RegionId} from "stores/widgets";
+import {SafeNumericInput} from "components/Shared";
+import {TaskProgressDialogComponent} from "components/Dialogs";
+import {Point2D} from "models";
+import "./PVGeneratorComponent.scss";
+
+export enum PVGeneratorComponentTabs {
+    PV_IMAGE,
+    PREVIEW_CUBE
+}
+
+@observer
+export class PVGeneratorComponent extends React.Component<WidgetProps> {
+    @observable selectedTabId: TabId = PVGeneratorComponentTabs.PV_IMAGE;
+
+    @action private setSelectedTab = (tab: TabId) => {
+        this.selectedTabId = tab;
+    };
+
+    public static get WIDGET_CONFIG(): DefaultWidgetConfig {
+        return {
+            id: "pv-generator",
+            type: "pv-generator",
+            minWidth: 350,
+            minHeight: 200,
+            defaultWidth: 500,
+            defaultHeight: 300,
+            title: "PV Generator",
+            isCloseable: true,
+            helpType: HelpType.PV_GENERATOR
+        };
+    }
+
+    @observable width: number;
+    @observable height: number;
+
+    @computed get widgetStore(): PVGeneratorWidgetStore {
+        const widgetsStore = WidgetsStore.Instance;
+        if (widgetsStore.pvGeneratorWidgets) {
+            const widgetStore = widgetsStore.pvGeneratorWidgets.get(this.props.id);
+            if (widgetStore) {
+                return widgetStore;
+            }
+        }
+        console.log("can't find store for widget");
+        return new PVGeneratorWidgetStore();
+    }
+
+    @computed get isLineIntersectedWithImage(): boolean {
+        if (this.widgetStore.effectiveRegion) {
+            const startPoint = this.widgetStore.effectiveRegion.controlPoints[0];
+            const endPoint = this.widgetStore.effectiveRegion.controlPoints[1];
+            const width = this.widgetStore.effectiveFrame.frameInfo.fileInfoExtended.width;
+            const height = this.widgetStore.effectiveFrame.frameInfo.fileInfoExtended.height;
+            const imageCorners: Point2D[] = [
+                {x: 0, y: 0},
+                {x: width - 1, y: 0},
+                {x: 0, y: height - 1},
+                {x: width - 1, y: height - 1}
```

Comment:
shall the positions be `-0.5, width/height - 0.5` because of the width of half pixel?

Reply:
done.